README.md
python-rfidtag
==============

Python packaging of tiki/touchatag reading code

This is a basic library for accessing tikitag (and touchatag) tag ids from
within python.

This is my initial packaging of the pytiki code by 
    * Younghun Kim
    * Thomas Schmid
    * Aaron Dunn

(as per : http://groups.google.com/group/tikitag/browse_frm/thread/c3aa678690800240# )
The original license for this code is GPL. This means that any program that
uses this code must also be GPL.  This fundamentally means you may not close
this source, nor may any application that uses this code be closed.  See the
license for exact wording and details.

All the files inside src/ are unmodified. Both the examples in examples/ are
modified slightly - dd.p is renamed to "basic.py", test.py has try...finally
added to ensure the tikitag reader is released cleanly.

Given this wraps a small piece of code using swig, it's plausible that this
could be adapted to support other dynamic languages as well.  If you do
this, please contribute back a patch.

Original code:
    * Younghun Kim
    * Thomas Schmid
    * Aaron Dunn

This packaging (setup.py, directory structure, etc): 
    * Michael Sparks

packaging contact: sparks.m@gmail.com
See also: http://groups.google.com/group/tikitag

http://groups.google.com/group/tikitag/browse_frm/thread/c3aa678690800240#

Usage
=====

Please see the example code to see usage. One caveat - due to using libusb,
your code MUST run as root - in order to have appropriate permissions to
open the tikitag device.

Again, if anyone has a patch or suggestion to avoid running as root, please
share.

---

Michael Sparks, February 2009
sparks.m@gmail.com

Who am I?
    * Michael Sparks - sparks.m@gmail.com . If I'm stepping on anyone's toes
      releasing this packaging, please let me know. (This readme will
      probably be replaced by something nicer, but this is a placemark
      before I release it and as per GPL I'm marking my changes including
      license file etc)



CHANGELOG
CHANGELOG

0.2 - Michael Sparks <sparks.m@gmail.com> 23 Sept 2012

    The codebase now supports touchtag readers as well as tikitag
    readers. 

    This forces two changes:

       * The name needs to change to rfidtag rather than tiki (for hopefully
         obvious reasons)
       * The examples need to change to match this obviously. (the kamaelia one
         has been simplified and made to work)
       * The touchatag readers have a different USB ID, and so that needs 
         including.

0.1 - Michael Sparks <sparks.m@gmail.com> 23 Sept 2012

    Transformed to use distutils for installation. Initial release.




Initial version retrieved from 
https://projects.nesl.ucla.edu/view/viewcvs.cgi/tikitag/trunk/pytiki/


examples/test.py
#!/usr/bin/python

from rfidtag import *

rfidtag_init()
rfidtag_config(5000,0)

while True:
    num = rfidtag_seek()
    if num == 0:
        print "No tag found"
    elif num == 1:
        print "Tag: ", rfidtag_getID1()
    elif num == 2:
        print "Tag 1/2", rfidtag_getID1()
        print 
        print "Tag 2/2", rfidtag_getID2()
    else:
        print "Nope", num

rfidtag_close()


src/tikitags.c
//sudo /etc/init.d/pcscd stop
//gcc -Wall -W -L/usr/lib/ -lusb tikitags.c -o tiki

#include <usb.h>
#include <stdio.h>

char INIT1[15] = { 0x6f, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x14, 0x00, 0x00, 0x08 };
char INIT2[15] = { 0x6f, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x14, 0x04, 0x00, 0x06 };
char INIT3[21] = { 0x6f, 0x0b, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF, 0x00, 0x00, 0x00, 0x06, 0xD4, 0x32, 0x05, 0x00, 0x00, 0x00 };
char INIT4[15] = { 0x6f, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xC0, 0x00, 0x00, 0x04 };
char status1[10] = { 0x65, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
char power_up[10] = { 0x62, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00 };
char power_down[10] = { 0x63, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
char abort_pipe[10] = { 0x72, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
char poll1[20] = { 0x6f, 0x0a, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00, 0x05, 0xd4, 0x60, 0x01, 0x01, 0x10 };
char get1[15] = { 0x6f, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xc0, 0x00, 0x00, 0x05 };
char recvbuf[64] = { 0x00 };            //full data hex array recieved from reader (includes header, full in specifed in ACR122 API, etc)
char id1[16] = { 0x00 };                //id of first tag found, [0]=reader found on, [1:8]=tag ID
char id2[16] = { 0x00 };                //id of second tag if two tags are found on one reader
unsigned long long id1_int = -1;
unsigned long long id2_int = -1;
int readerID = -1;
char count = 0;                         //insturction count, [6] of each header is the count. May not be necessary...
int numoftiki = 0;                      //number of tikitag readers
int seektime = 0;		//in ms
int looprate = 0;		//in ms max of 1000ms (1sec)
struct usb_device* tikidev[10];
struct usb_dev_handle* tiki[10];

 /*
  ** tiki_xchange(udp, *outp, out_len, *inp, in_len)
  **      send a command to the reader and gets the answer
  **      returns 0 if all goes well; -1 if not
  **LOCKED IN
  */
int
tiki_xchange (struct usb_dev_handle *udp, char *outp, int out_len, char *inp,
	      int in_len)
{
  int bytes_sent;
  int bytes_recv;

  count++;
  if ((bytes_sent =
       usb_bulk_write (udp, 0x02, (char *) outp, out_len, 10)) < 0)
    {
      if ((bytes_sent =
	   usb_bulk_write (udp, 0x02, (char *) outp, out_len, 10)) < 0)
	{
	  perror ("usb_bulk_write");
	}
    }
  if ((bytes_recv = usb_bulk_read (udp, 0x82, (char *) inp, in_len, 10)) < 0)
    {
      if ((bytes_recv =
	   usb_bulk_read (udp, 0x82, (char *) inp, in_len, 100)) < 0)
	{
	  perror ("usb_bulk_read");
	  return (-1);
	}
    }

  return bytes_recv;
}

 /*
  **
  **Power up and send Init Commands from working python code
  **        Init3[]-Immediate result of polling command
  **LOCKED IN
  */

int
tiki_setup (struct usb_dev_handle *udp)
{
  power_up[6] = count;
  if (tiki_xchange
      (udp, power_up, sizeof (power_up), recvbuf, sizeof (recvbuf)) < 0)
    {
      perror ("xchange");
      return (-1);
    }

  INIT1[6] = count;
  if (tiki_xchange (udp, INIT1, sizeof (INIT1), recvbuf, sizeof (recvbuf)) <
      0)
    {
      perror ("xchange");
      return (-1);
    }
  INIT2[6] = count;
  if (tiki_xchange (udp, INIT2, sizeof (INIT2), recvbuf, sizeof (recvbuf)) <
      0)
    {
      perror ("xchange");
      return (-1);
    }
  INIT3[6] = count;
  if (tiki_xchange (udp, INIT3, sizeof (INIT3), recvbuf, sizeof (recvbuf)) <
      0)
    {
      perror ("xchange");
      return (-1);
    }
  INIT4[6] = count;
  if (tiki_xchange (udp, INIT4, sizeof (INIT4), recvbuf, sizeof (recvbuf)) <
      0)
    {
      perror ("xchange");
      return (-1);
    }
  return (0);
}

unsigned long long
tiki_getID1 ()
{
  return id1_int;
}

unsigned long long
tiki_getID2 ()
{
  return id2_int;
}

int
tiki_getReaderID ()
{
  return readerID;
}

/*
**
**Tikitag Initalization Sequence
**  Returns numbe rof devices found
**LOCKED IN
*/
int
tiki_init (void)
{
  struct usb_bus *busses, *bus;
  struct usb_device *dev;

  int i = 0;

  usb_init ();
  usb_find_busses ();
  usb_find_devices ();

  busses = usb_get_busses ();
  for (bus = busses; bus; bus = bus->next)
    {
      for (dev = bus->devices; dev; dev = dev->next)
	{
	  if (dev->descriptor.idVendor == 0x072f &&	/* ACS */
	      dev->descriptor.idProduct == 0x90cc)	/* ACR122 */
	    {
	      printf ("Found tikitag reader %d!\n", numoftiki);
	      tikidev[numoftiki] = dev;
	      numoftiki++;
	    }
	  //printf("works\n");
	}
    }
  if (numoftiki == 0)
    {
      printf ("Didn't Find tikitag reader...\n");
      exit (-1);
    }
  for (i = 0; i < numoftiki; i++)
    {

      if ((tiki[i] = usb_open (tikidev[i])) == NULL)
	{
	  perror ("usb_open");
	  exit (-1);
	}
       /*      
      if(usb_claim_interface(tiki[i], 0)<0)
       {
	  perror("usb_claim_interface");
	  exit(-1);
       }
       */
      //printf("Open[%d]!\n",i+1);
    }
  for (i = 0; i < numoftiki; i++)
    {
      if ((tiki_setup (tiki[i])) < 0)
	{
	  perror ("tiki_init");
	  exit(-1);
	}
    }
  return (numoftiki);
}

 /*
  **
  **Decode Input from tag to just return UID
  **       Reader seems to be able to read a max of two at a time
  **LOCKED IN
  */
int
tiki_decode (char *inp, int reader)
{
  int numtags = inp[12];
  int idlen1 = inp[19];
  int idlen2 = inp[33];
  int i;

  if (numtags == 0)
    {
      //printf("Reader[%d]:No tags present\n\n", reader+1);
      return (-1);
    }
  readerID = reader;
  if (numtags >= 1)		//One Tag
    {
      //id1[0] = reader + 1;
      for (i = 0; i < 8; i++)	//Load ID
	{
	  id1[i] = inp[i + 18];
	}
      //printf("Reader[%d]:ID = ", reader+1);                //Print ID
      id1_int = 0;
      for (i = 0; i < idlen1 + 1; i++)
	{
	  id1_int += id1[i] << (i * 8);
	}
      //printf("%u", id1_int);
      //printf("\n\n");
    }
  if (numtags == 2)		//Two Tags
    {
      id2[0] = reader + 1;
      for (i = 0; i < 8; i++)	//Load ID2
	{
	  id2[i] = inp[i + 32];
	}
      //printf("Reader[%d]:ID2 = ", reader+1);
      for (i = 0; i < idlen2 + 1; i++)
	{
	  id2_int += id2[i] << (i * 8);
	}
      //printf("%u", id2_int);
      //printf("\n\n");
    }
  return numtags;
}

 /*
  **
  **tiki_config config seektime and looprate
  **
  **LOCKED IN
  */
int
tiki_config (int time, int rate)
{
  seektime = time;

  if (rate < 1000)
    looprate = rate;
  if (rate > 1000)
    looprate = 1000;
  return (0);
}

 /*
  **
  **tiki close - power down reader and close usb communication
  **    
  **LOCKED IN
  */
int
tiki_close (void)
{
  int i = 0;
  for (i = 0; i < numoftiki; i++)
    {
      if (tiki_xchange
	  (tiki[i], power_down, sizeof (power_down), recvbuf,
	   sizeof (recvbuf)) < 0)
	{
	  perror ("xchange");
	}
      //printf("closing[%d]\n\n", i+1);
      if (usb_close (tiki[i]) < 0)
	{
	  perror ("usb_close");
	  return (-1);
	}
    }
  return (0);
}

 /*
  **
  **tiki_seek - searches for time set in tiki_config (time) at the rate specfied in tiki_config (rate)
  **             rate is how often program checks for a tag, once tag is found store in global variable
  **             ID1 is two are found after a poll they are stored in ID1 and ID2
  **
  **LOCKED IN
  */
int
tiki_seek (void)
{				//seektime/(looprate+65ms)=iterations (65ms or 95ms is roughly time to run through one iteration)
  int t = 0;
  int i = 0;
  int recvb;
  int numtags;
  while (t < (seektime / (looprate + 95)))
    {

      for (i = 0; i < numoftiki; i++)
	{
          //printf("haha %d\n", count);
	  poll1[6] = count;
	  if ((recvb =
	       tiki_xchange (tiki[i], poll1, sizeof (poll1), recvbuf,
			     sizeof (recvbuf))) < 0)
	    {
	      perror ("xchange");
	      return (-1);
	    }
	  get1[6] = count;
	  get1[14] = recvbuf[11];
	  if ((recvb =
	       tiki_xchange (tiki[i], get1, sizeof (get1), recvbuf,
			     sizeof (recvbuf))) < 0)
	    {
	      perror ("xchange");
	    }
	  //Add debug mode for Printing
	  numtags = tiki_decode (recvbuf, i);
	  if (numtags > 0)	//check num of tags
		  return numtags;		//return num of tags
	}
      t++;
      usleep (looprate * 1000);	//convert ms to us
    }
  return 0;
}


examples/kamaelia-tiki.py
#! /usr/bin/python

import pytiki as ti
import Axon
import time

class TikiTagDownUp(Axon.ThreadedComponent.threadedcomponent):
    def main(self):
        print "Woo"
        tagstatus = {}
        seentag = {}
        try:
            ti.tiki_init()
            ti.tiki_config(196,0)
            while 1:
                num = ti.tiki_seek()
                if num == 0:
                    for i in tagstatus:
#                        self.send(("TAGUP", i), "outbox")
                        print time.time(), "TAG UP  : ", i
                    tagstatus = {}
                elif num == 1:
                    raw_tag_id = ti.tiki_getID1()
                    if seentag.get(raw_tag_id, False):
                        if tagstatus.get(raw_tag_id, "up") != "down":
                            print time.time(), "TAG DOWN: ", raw_tag_id
#                            self.send(("TAGDOWN", raw_tag_id), "outbox")
#                            print time.time(), "TAG DOWN: ", raw_tag_id
                            tagstatus[raw_tag_id] = "down"
                    else:
#                        self.send(("TAGDOWN", raw_tag_id), "outbox")
                        print time.time(), "TAG DOWN:", raw_tag_id
                        seentag[raw_tag_id] = True
                        tagstatus[raw_tag_id] = "down"

        finally:
            ti.tiki_close()


from Kamaelia.Chassis.Pipeline import Pipeline
from Kamaelia.Util.Console import ConsoleEchoer

Pipeline( TikiTagDownUp(), ConsoleEchoer(use_repr=True) ).run()




src/rfidtag.c
//sudo /etc/init.d/pcscd stop
//gcc -Wall -W -L/usr/lib/ -lusb rfidtag.c -o rfidtag

#include <usb.h>
#include <stdio.h>

char INIT1[15] = { 0x6f, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x14, 0x00, 0x00, 0x08 };
char INIT2[15] = { 0x6f, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x14, 0x04, 0x00, 0x06 };
char INIT3[21] = { 0x6f, 0x0b, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF, 0x00, 0x00, 0x00, 0x06, 0xD4, 0x32, 0x05, 0x00, 0x00, 0x00 };
char INIT4[15] = { 0x6f, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xC0, 0x00, 0x00, 0x04 };
char status1[10] = { 0x65, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
char power_up[10] = { 0x62, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00 };
char power_down[10] = { 0x63, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
char abort_pipe[10] = { 0x72, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
char poll1[20] = { 0x6f, 0x0a, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00, 0x05, 0xd4, 0x60, 0x01, 0x01, 0x10 };
char get1[15] = { 0x6f, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xc0, 0x00, 0x00, 0x05 };
char recvbuf[64] = { 0x00 };            //full data hex array recieved from reader (includes header, full in specifed in ACR122 API, etc)
char id1[16] = { 0x00 };                //id of first tag found, [0]=reader found on, [1:8]=tag ID
char id2[16] = { 0x00 };                //id of second tag if two tags are found on one reader
unsigned long long id1_int = -1;
unsigned long long id2_int = -1;
int readerID = -1;
char count = 0;                         //insturction count, [6] of each header is the count. May not be necessary...
int numtagreaders = 0;                      //number of rfidtag readers
int seektime = 0;		//in ms
int looprate = 0;		//in ms max of 1000ms (1sec)
struct usb_device* rfid_devices[10];
struct usb_dev_handle* rfid_dev[10];

 /*
  ** rfidtag_xchange(udp, *outp, out_len, *inp, in_len)
  **      send a command to the reader and gets the answer
  **      returns 0 if all goes well; -1 if not
  **LOCKED IN
  */
int
rfidtag_xchange (struct usb_dev_handle *udp, char *outp, int out_len, char *inp,
	      int in_len)
{
  int bytes_sent;
  int bytes_recv;

  count++;
  if ((bytes_sent =
       usb_bulk_write (udp, 0x02, (char *) outp, out_len, 10)) < 0)
    {
      if ((bytes_sent =
	   usb_bulk_write (udp, 0x02, (char *) outp, out_len, 10)) < 0)
	{
	  perror ("usb_bulk_write");
	}
    }
  if ((bytes_recv = usb_bulk_read (udp, 0x82, (char *) inp, in_len, 10)) < 0)
    {
      if ((bytes_recv =
	   usb_bulk_read (udp, 0x82, (char *) inp, in_len, 100)) < 0)
	{
	  perror ("usb_bulk_read");
	  return (-1);
	}
    }

  return bytes_recv;
}

 /*
  **
  **Power up and send Init Commands from working python code
  **        Init3[]-Immediate result of polling command
  **LOCKED IN
  */

int
rfidtag_setup (struct usb_dev_handle *udp)
{
  power_up[6] = count;
  if (rfidtag_xchange
      (udp, power_up, sizeof (power_up), recvbuf, sizeof (recvbuf)) < 0)
    {
      perror ("xchange");
      return (-1);
    }

  INIT1[6] = count;
  if (rfidtag_xchange (udp, INIT1, sizeof (INIT1), recvbuf, sizeof (recvbuf)) <
      0)
    {
      perror ("xchange");
      return (-1);
    }
  INIT2[6] = count;
  if (rfidtag_xchange (udp, INIT2, sizeof (INIT2), recvbuf, sizeof (recvbuf)) <
      0)
    {
      perror ("xchange");
      return (-1);
    }
  INIT3[6] = count;
  if (rfidtag_xchange (udp, INIT3, sizeof (INIT3), recvbuf, sizeof (recvbuf)) <
      0)
    {
      perror ("xchange");
      return (-1);
    }
  INIT4[6] = count;
  if (rfidtag_xchange (udp, INIT4, sizeof (INIT4), recvbuf, sizeof (recvbuf)) <
      0)
    {
      perror ("xchange");
      return (-1);
    }
  return (0);
}

unsigned long long
rfidtag_getID1 ()
{
  return id1_int;
}

unsigned long long
rfidtag_getID2 ()
{
  return id2_int;
}

int
rfidtag_getReaderID ()
{
  return readerID;
}

/*
**
**Tikitag Initalization Sequence
**  Returns numbe rof devices found
**LOCKED IN
*/
int
rfidtag_init (void)
{
  struct usb_bus *busses, *bus;
  struct usb_device *dev;

  int i = 0;

  usb_init ();
  usb_find_busses ();
  usb_find_devices ();

  busses = usb_get_busses ();
  for (bus = busses; bus; bus = bus->next)
    {
      for (dev = bus->devices; dev; dev = dev->next)
	{
	  if (dev->descriptor.idVendor == 0x072f) /* ACS */
	  { 
              if (dev->descriptor.idProduct == 0x90cc)	/* ACR122 */
	      {
	          printf ("Found tikitag reader %d!\n", numtagreaders);
	          rfid_devices[numtagreaders] = dev;
	          numtagreaders++;
	      }
              if (dev->descriptor.idProduct == 0x2200)	/* ACR122 */
	      {
	          printf ("Found touchatag reader %d!\n", numtagreaders);
	          rfid_devices[numtagreaders] = dev;
	          numtagreaders++;
	      }
          }
	  //printf("works\n");
	}
    }
  if (numtagreaders == 0)
    {
      printf ("Didn't Find rfidtag reader...\n");
      exit (-1);
    }
  for (i = 0; i < numtagreaders; i++)
    {

      if ((rfid_dev[i] = usb_open (rfid_devices[i])) == NULL)
	{
	  perror ("usb_open");
	  exit (-1);
	}
       /*      
      if(usb_claim_interface(rfid_dev[i], 0)<0)
       {
	  perror("usb_claim_interface");
	  exit(-1);
       }
       */
      //printf("Open[%d]!\n",i+1);
    }
  for (i = 0; i < numtagreaders; i++)
    {
      if ((rfidtag_setup (rfid_dev[i])) < 0)
	{
	  perror ("rfidtag_init");
	  exit(-1);
	}
    }
  return (numtagreaders);
}

 /*
  **
  **Decode Input from tag to just return UID
  **       Reader seems to be able to read a max of two at a time
  **LOCKED IN
  */
int
rfidtag_decode (char *inp, int reader)
{
  int numtags = inp[12];
  int idlen1 = inp[19];
  int idlen2 = inp[33];
  int i;

  if (numtags == 0)
    {
      //printf("Reader[%d]:No tags present\n\n", reader+1);
      return (-1);
    }
  readerID = reader;
  if (numtags >= 1)		//One Tag
    {
      //id1[0] = reader + 1;
      for (i = 0; i < 8; i++)	//Load ID
	{
	  id1[i] = inp[i + 18];
	}
      //printf("Reader[%d]:ID = ", reader+1);                //Print ID
      id1_int = 0;
      for (i = 0; i < idlen1 + 1; i++)
	{
	  id1_int += id1[i] << (i * 8);
	}
      //printf("%u", id1_int);
      //printf("\n\n");
    }
  if (numtags == 2)		//Two Tags
    {
      id2[0] = reader + 1;
      for (i = 0; i < 8; i++)	//Load ID2
	{
	  id2[i] = inp[i + 32];
	}
      //printf("Reader[%d]:ID2 = ", reader+1);
      for (i = 0; i < idlen2 + 1; i++)
	{
	  id2_int += id2[i] << (i * 8);
	}
      //printf("%u", id2_int);
      //printf("\n\n");
    }
  return numtags;
}

 /*
  **
  **rfidtag_config config seektime and looprate
  **
  **LOCKED IN
  */
int
rfidtag_config (int time, int rate)
{
  seektime = time;

  if (rate < 1000)
    looprate = rate;
  if (rate > 1000)
    looprate = 1000;
  return (0);
}

 /*
  **
  **rfidtag close - power down reader and close usb communication
  **    
  **LOCKED IN
  */
int
rfidtag_close (void)
{
  int i = 0;
  for (i = 0; i < numtagreaders; i++)
    {
      if (rfidtag_xchange
	  (rfid_dev[i], power_down, sizeof (power_down), recvbuf,
	   sizeof (recvbuf)) < 0)
	{
	  perror ("xchange");
	}
      //printf("closing[%d]\n\n", i+1);
      if (usb_close (rfid_dev[i]) < 0)
	{
	  perror ("usb_close");
	  return (-1);
	}
    }
  return (0);
}

 /*
  **
  **rfidtag_seek - searches for time set in rfidtag_config (time) at the rate specfied in rfidtag_config (rate)
  **             rate is how often program checks for a tag, once tag is found store in global variable
  **             ID1 is two are found after a poll they are stored in ID1 and ID2
  **
  **LOCKED IN
  */
int
rfidtag_seek (void)
{				//seektime/(looprate+65ms)=iterations (65ms or 95ms is roughly time to run through one iteration)
  int t = 0;
  int i = 0;
  int recvb;
  int numtags;
  while (t < (seektime / (looprate + 95)))
    {

      for (i = 0; i < numtagreaders; i++)
	{
          //printf("haha %d\n", count);
	  poll1[6] = count;
	  if ((recvb =
	       rfidtag_xchange (rfid_dev[i], poll1, sizeof (poll1), recvbuf,
			     sizeof (recvbuf))) < 0)
	    {
	      perror ("xchange");
	      return (-1);
	    }
	  get1[6] = count;
	  get1[14] = recvbuf[11];
	  if ((recvb =
	       rfidtag_xchange (rfid_dev[i], get1, sizeof (get1), recvbuf,
			     sizeof (recvbuf))) < 0)
	    {
	      perror ("xchange");
	    }
	  //Add debug mode for Printing
	  numtags = rfidtag_decode (recvbuf, i);
	  if (numtags > 0)	//check num of tags
		  return numtags;		//return num of tags
	}
      t++;
      usleep (looprate * 1000);	//convert ms to us
    }
  return 0;
}
